cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -980,6 +980,36 @@ struct CastFunctor<O, StringType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Timestamp
+
+template <>
+struct CastFunctor<TimestampType, StringType> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    //     const auto& out_type = checked_cast<const TimestampType&>(*output->type);
+    using out_type = TimestampType::c_type;
+
+    StringArray input_array(input.Copy());
```

Comment:
Why does it need to copy?

Reply:
This is a shallow copy (it doesn't copy the underlying data, just duplicate the containers). But good question. I took it from the other casting implementations. Perhaps constructing an Array may mutate the underlying ArrayData and we don't want that? @wesm

Reply:
NB: the copy is shallow / the memory buffers are not copied (I am not sure why it is needed though)

Reply:
Yes.

Reply:
As for the copy, actually it's simple: `Array` expects a `shared_ptr<ArrayData>`, and we have a bare reference here, from which we cannot safely devise a `shared_ptr` unless we copy it...